Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/presencetracing/organizer/warn/duration/TraceLocationWarnDurationViewModel.kt

```diff
@@ -0,0 +1,107 @@
+import kotlinx.coroutines.flow.MutableStateFlow
+import org.joda.time.Duration
+import org.joda.time.Instant
+import org.joda.time.LocalDate
+import org.joda.time.LocalTime
+
+class TraceLocationWarnDurationViewModel @AssistedInject constructor(
+    @Assisted private val traceLocation: TraceLocation,
+    dispatcherProvider: DispatcherProvider,
+    private val timeStamper: TimeStamper,
+) : CWAViewModel(dispatcherProvider = dispatcherProvider) {
+
+    private val uiState = MutableStateFlow(UiState())
+    val state: LiveData<UiState> = uiState.asLiveData2()
+    val selectedDate: LocalDate get() = uiState.value.localDate
+    val selectedTime: LocalTime get() = uiState.value.localTime
+    val selectedDuration: Duration get() = uiState.value.duration
+
+    init {
+        updateUiState()
+    }
+
+    fun dateChanged(localDate: LocalDate, localTime: LocalTime) {
+        val resultTime = if (localDate == LocalDate.now() &&
+            localTime.isAfter(LocalTime.now())
+        ) LocalTime.now() else localTime
+        uiState.apply { value = value.copy(localDate = localDate, localTime = resultTime) }
+    }
+
+    fun durationChanged(duration: Duration) {
+        uiState.apply {
+            value = value.copy(
+                duration = Duration.standardMinutes(duration.standardMinutes.coerceIn(15, 1425))
+            )
+        }
+    }
+
+    private fun updateUiState() {
+        uiState.apply {
+            value = value.copy(
+                description = traceLocation.description,
+                startDateTime = traceLocation.startDate,
+                endDateTime = traceLocation.endDate
+            )
+        }
+
+        if (traceLocation.startDate != null &&
+            traceLocation.startDate != Instant.EPOCH &&
+            !traceLocation.isBeforeStartTime(timeStamper.nowUTC)
+        ) {
+            uiState.apply {
+                value = value.copy(
+                    localDate = traceLocation.startDate.toDateTime().toLocalDate(),
+                    localTime = traceLocation.startDate.toDateTime().toLocalTime()
+                )
+            }
+        }
+
+        val duration = if (traceLocation.endDate != null && traceLocation.endDate != Instant.EPOCH) {
+            getNearestFifteen(Duration(traceLocation.startDate, traceLocation.endDate).standardMinutes)
+        } else if (traceLocation.defaultCheckInLengthInMinutes != null &&
+            traceLocation.defaultCheckInLengthInMinutes > 0
+        ) {
+            getNearestFifteen(traceLocation.defaultCheckInLengthInMinutes.toLong())
+        } else {
+            Duration.standardHours(2)
+        }
+        durationChanged(duration)
+    }
+
+    private fun getNearestFifteen(durationMinutes: Long): Duration {
+        val minDif = durationMinutes % 15
+        val maxDif = 15 - (durationMinutes % 15)
+        return Duration.standardMinutes(if (minDif < maxDif) durationMinutes - minDif else durationMinutes + maxDif)
+    }
```

Comment:
I wonder if there is more idiomatic way to get nearest to 15 value 🤔
Any suggestions?

Reply:
I was trying Google to find suggestions ('kotlin rounding'). From all results, yours looks cleanest and leanest! If none of your colleagues has a better solution: chapeau!

Reply:
Would that work? Some unit tests would be helpful to verify its correctness 
```kotlin
fun roundToNearest15(number: Int): Int {
    return round(number.toFloat() / 15).toInt() * 15
}
```

Some examples
```console
32 rounded to 30
55 rounded to 60
51 rounded to 45
60 rounded to 60
85 rounded to 90
9 rounded to 15
105 rounded to 105
5 rounded to 0
10 rounded to 15
15 rounded to 15
20 rounded to 15
24 rounded to 30
35 rounded to 30
40 rounded to 45
45 rounded to 45
50 rounded to 45

```